### DISCUSSION TITLE
Different configurations for main and test in the detekt task possible?

### DISCUSSION THREAD

**Discussion by mrclrchtr** [detekt/detekt#5499] [Q&A]
Hello,

I have configured a multi module project with gradle as follows:

buildSrc/src/main/kotlin/detekt-conventions.gradle.kts:
```
val baselineFiles = file("${rootProject.projectDir}/detekt/baseline.xml")
val configFiles = files("${rootProject.projectDir}/detekt/detekt.yml", "${rootProject.projectDir}/detekt/detekt-test.yml")

detekt {
  config = configFiles
  buildUponDefaultConfig = true
  baseline = baselineFiles
  parallel = true
}

tasks {
  withType<Detekt>().configureEach {
    reports {
      xml.required.set(true)
      html.required.set(true)
      md.required.set(true)
    }
  }
}

val detektBaselineMultiModule by tasks.registering(DetektCreateBaselineTask::class) {
  description = "Overrides current baseline."
  buildUponDefaultConfig.set(true)
  ignoreFailures.set(true)
  parallel.set(true)
  setSource(files(rootProject.projectDir))
  config.setFrom(configFiles)
  baseline.set(baselineFiles)
  include("**/*.kt")
  include("**/*.kts")
  exclude("**/resources/**")
  exclude("**/build/**")
}
```
(used in the modules)

detektBaselineMultiModule generates me the correct baseline in which violations from test are also ignored.

The Problem:
If i disable a check in the detekt.yml but not in the detekt-test.yml, the check in the main sources still applies on the 'detekt' task. Probably because detekt-test.yml is more specific? (the baseline is working correctly)

I would have expected detekt-test.yml to apply only to the test resources?

One of my solutions was to use the 'detektMain' and 'detektTest' task separately instead of 'detekt'. But they didn't seem to use the baseline at all. I do not understand that yet. Do I have to configure this differently?

Thank you for your help!

**Comment by BraisGabin** [comment]
detekt doesn't allow right now to have multiple configurations. Well... I imagine that if you configure two different tasks you should be able to do that but that's work that you should do on your gradle. It's not provided out of the box.

If you provides two different configuration files the entires of the second one overwrite the ones in the first one

**Comment by cortinico** [comment]
> I imagine that if you configure two different tasks you should be able to do that but that's work that you should do on your gradle

I'd echo what @BraisGabin already said. You can create your own custom tasks with different configs.

On top of this, you can use the `excludes` in the config file, to let some rule runs only on `src/main` or `src/test` if that is your file structure.

**Comment by mrclrchtr** [comment]
Ok, thank you very much for your help. Is it also the case that both are always used, even if I use detektMain and detektTest? Or is the detekt-main.yml and detekt-test.yml used respectively?

I am currently trying to solve it with these two tasks. But then type resolution is activated (which I think is good). But I don't manage to teach type resolution to my baseline task in the multi module project... do you have an idea what is missing?

```
val baselineFiles = file("${rootProject.projectDir}/detekt/baseline.xml")
val configFiles = files("${rootProject.projectDir}/detekt/detekt.yml")

val detektBaselineMultiModuleMain by tasks.registering(DetektCreateBaselineTask::class) {
  jvmTarget = "11"
  buildUponDefaultConfig.set(true)
  ignoreFailures.set(true)
  parallel.set(true)
  setSource(files(rootProject.projectDir))
  config.setFrom(configFiles)
  baseline.set(baselineFiles)
  include("**/*.kt")
  include("**/*.kts")
  exclude("**/resources/**")
  exclude("**/build/**")
}
```

**Reply by cortinico** [reply]
> Ok, thank you very much for your help. Is it also the case that both are always used, even if I use detektMain and detektTest? Or is the detekt-main.yml and detekt-test.yml used respectively?
> 
> I am currently trying to solve it with these two tasks. But then type resolution is activated (which I think is good). But I don't manage to teach type resolution to my baseline task in the multi module project... do you have an idea what is missing?

If you're looking into using Type Resolution, I would suggest you use a single config file + use the excludes based on path.

**Reply by mrclrchtr** [reply]
Yes, this works as intended, but how can I create a single baseline file with type resolution in a multi module project?

**Reply by cortinico** [reply]
Creating a single baseline is not how Detekt is designed to work in a multi-module project, sadly (see https://github.com/detekt/detekt/issues/1589)
